x/auth/ante/unordered.go

```diff
@@ -26,49 +45,195 @@ var _ sdk.AnteDecorator = (*UnorderedTxDecorator)(nil)
+	// This is done to avoid sha256 computation in simulation mode
+	if sdkCtx.ExecMode() == sdk.ExecModeSimulate {
+		return nil
+	}
+
+	// calculate the tx hash
+	txHash, err := TxIdentifier(uint64(timeoutTimestamp.Unix()), sdkTx)
+	if err != nil {
+		return err
+	}
 
 	// check for duplicates
 	if d.txManager.Contains(txHash) {
-		return ctx, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "tx %X is duplicated")
+		return errorsmod.Wrap(
+			sdkerrors.ErrInvalidRequest,
+			"tx %X is duplicated",
+		)
 	}
-
-	if ctx.ExecMode() == sdk.ExecModeFinalize {
+	if sdkCtx.ExecMode() == sdk.ExecModeFinalize {
 		// a new tx included in the block, add the hash to the unordered tx manager
-		d.txManager.Add(txHash, ttl)
+		d.txManager.Add(txHash, timeoutTimestamp)
+	}
+
+	return nil
+}
+
+// TxIdentifier returns a unique identifier for a transaction that is intended to be unordered.
+func TxIdentifier(timeout uint64, tx sdk.Tx) ([32]byte, error) {
+	sigTx, ok := tx.(authsigning.Tx)
+	if !ok {
+		return [32]byte{}, errorsmod.Wrap(sdkerrors.ErrTxDecode, "invalid transaction type")
+	}
+
+	if sigTx.GetFee().IsZero() {
+		return [32]byte{}, errorsmod.Wrap(
+			sdkerrors.ErrInvalidRequest,
+			"unordered transaction must have a fee",
+		)
+	}
+
+	buf := bufPool.Get().(*bytes.Buffer)
+	// Make sure to reset the buffer
+	buf.Reset()
+	defer bufPool.Put(buf)
+
+	// Add signatures to the transaction identifier
+	signatures, err := sigTx.GetSignaturesV2()
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	for _, sig := range signatures {
+		if err := addSignatures(sig.Data, buf); err != nil {
+			return [32]byte{}, err
+		}
+	}
+
+	// Use the buffer
+	for _, msg := range tx.GetMsgs() {
+		// loop through the messages and write them to the buffer
+		// encoding the msg to bytes makes it deterministic within the state machine.
+		// Malleability is not a concern here because the state machine will encode the transaction deterministically.
+		bz, err := proto.Marshal(msg)
+		if err != nil {
+			return [32]byte{}, errorsmod.Wrap(
+				sdkerrors.ErrInvalidRequest,
+				"failed to marshal message",
+			)
+		}
+
+		if _, err := buf.Write(bz); err != nil {
+			return [32]byte{}, errorsmod.Wrap(
+				sdkerrors.ErrInvalidRequest,
+				"failed to write message to buffer",
+			)
+		}
+	}
+
+	// write the timeout height to the buffer
+	if err := binary.Write(buf, binary.LittleEndian, timeout); err != nil {
+		return [32]byte{}, errorsmod.Wrap(
+			sdkerrors.ErrInvalidRequest,
+			"failed to write timeout_height to buffer",
+		)
+	}
+
+	// write gas to the buffer
+	if err := binary.Write(buf, binary.LittleEndian, sigTx.GetGas()); err != nil {
+		return [32]byte{}, errorsmod.Wrap(
+			sdkerrors.ErrInvalidRequest,
+			"failed to write unordered to buffer",
+		)
+	}
+
+	txHash := sha256.Sum256(buf.Bytes())
+
+	// Return the Buffer to the pool
+	return txHash, nil
+}
+
+func addSignatures(sig signing.SignatureData, buf *bytes.Buffer) error {
```

Comment:
this is scary because it recursive signning

Reply:
the `SigVerificationDecorator` will come before unordered tx decorator, so at this point we will know these are valid signatures and aren't just bogus sigs to bog the chain